From 7af228152535ce1b34cae4c92b1c90f57a226d28 Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@foss.st.com>
Date: Tue, 21 May 2024 14:08:11 +0200
Subject: [PATCH 1133/1141] Revert "net: stmmac: ethtool: Fixed calltrace
 caused by unbalanced disable_irq_wake calls"

Need to revert this commit, which introduce regression on WoL usecases:
(when enable WoL with "ethtool -s end0 wol g")
(BZ: https://intbugzilla.st.com/show_bug.cgi?id=181557)

E/TC:0   stm32_serc_handle_ilac:134 SERC exceptions [63:32]: 0x20000000
E/TC:0   stm32_serc_handle_ilac:140 SERC exception ID: 61
[   53.699524] stm32-dwmac 482d0000.eth2 end0: FPE workqueue stop
[   53.700009] stm32-dwmac 482c0000.eth1 end1: FPE workqueue stop
[   53.792329] Disabling non-boot CPUs ...
[   53.793982] psci: CPU1 killed (polled 0 ms)
[   53.795036] optee: no desc for optee IT:1
[   53.795333] Enabling non-boot CPUs ...
[   53.803275] Detected VIPT I-cache on CPU1
[   53.803330] cacheinfo: Unable to detect cache hierarchy for CPU 1
[   53.803377] CPU1: Booted secondary processor 0x0000000001 [0x411fd040]
[   53.803780] CPU1 is up
[   53.892352] Internal error: synchronous external abort: 0000000096000210 [#1] PREEMPT SMP
[   53.892367] Modules linked in: ucsi_stm32g0 typec_ucsi typec cfg80211 rfkill rpmsg_ctrl rpmsg_char usb_f_ncm u_ether stm32_dcmipp crct10dif_ce hantro_vpu v4l6
[   53.892524] CPU: 1 PID: 9 Comm: kworker/u4:0 Tainted: G           O       6.1.82 #1
[   53.892534] Hardware name: STMicroelectronics STM32MP257F-EV1 Evaluation Board (DT)
[   53.892540] Workqueue: events_power_efficient phy_state_machine
[   53.892560] pstate: 60000005 (nZCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   53.892571] pc : stmmac_mdio_read+0xa8/0x280
[   53.892584] lr : stmmac_mdio_read+0x94/0x280
[   53.892593] sp : ffff800009adbc50
[   53.892597] x29: ffff800009adbc50 x28: 0000000000000000 x27: 0000000000000000
[   53.892612] x26: ffff00008000c000 x25: ffff0000800ee010 x24: 0000000c81d81afc
[   53.892627] x23: 0000000000000200 x22: 0000000000000000 x21: 0000000000000204
[   53.892641] x20: 000000000020040d x19: ffff000083c80980 x18: 0000000000000008
[   53.892655] x17: 000000040044ffff x16: 00000000000000b3 x15: 0000000000000040
[   53.892670] x14: 0000000000000060 x13: 0000000000000060 x12: 0000000000004a7d
[   53.892684] x11: 0000000000000001 x10: 0000000000000a60 x9 : ffff800009adbd60
[   53.892698] x8 : ffff0000800b0ac0 x7 : ffff000083d2e080 x6 : 00000000a25e68a4
[   53.892713] x5 : 00ffffffffffffff x4 : 00359bc18b000000 x3 : 0000000000000018
[   53.892727] x2 : 0000000000003e8a x1 : ffff80000a120200 x0 : 0000000c813f847c
[   53.892742] Call trace:
[   53.892747]  stmmac_mdio_read+0xa8/0x280
[   53.892759]  mdiobus_read+0x40/0x80
[   53.892768]  genphy_update_link+0x24/0xf4
[   53.892781]  genphy_read_status+0x24/0x110
[   53.892789]  rtlgen_read_status+0x18/0x3c
[   53.892798]  phy_check_link_status+0x38/0xbc
[   53.892808]  phy_state_machine+0x19c/0x270
[   53.892818]  process_one_work+0x1d4/0x330
[   53.892829]  worker_thread+0x6c/0x430
[   53.892839]  kthread+0x108/0x10c
[   53.892849]  ret_from_fork+0x10/0x20
[   53.892864] Code: 2a1703f7 911a0318 f9405261 8b170021 (b9400021)
[   53.892871] ---[ end trace 0000000000000000 ]---
[   54.897172] stm32-pcie 48400000.pcie: Phy link never came up
[   54.938647] dwmac4: Master AXI performs any burst length
[   54.938680] stm32-dwmac 482c0000.eth1 end1: No Safety Features support found
[   55.138723] stm32-dwmac 482c0000.eth1 end1: IEEE 1588-2008 Advanced Timestamp supported
[   55.138944] stm32-dwmac 482c0000.eth1 end1: FPE workqueue start
[   55.138996] stm32-dwmac 482c0000.eth1 end1: configuring for phy/rgmii-id link mode
[   56.324602] stm32-dwmac 482d0000.eth2: Failed to reset the dma
[   56.324611] stm32-dwmac 482d0000.eth2 end0: stmmac_hw_setup: DMA engine initialization failed
[   56.324661] stm32-dwmac 482d0000.eth2 end0: Link is Down
[   56.327401] usb usb1: root hub lost power or was reset
[   56.620458] onboard-usb-hub 1-1: reset high-speed USB device number 2 using ehci-platform
[   57.219261] OOM killer enabled.
[   57.222295] Restarting tasks ... done.
[   57.246411] random: crng reseeded on system resumption
[   57.262436] PM: suspend exit


This reverts commit f1ec1b6c5c469754c99abe2606147ccaa280e794.
Signed-off-by: Christophe Roullier <christophe.roullier@foss.st.com>
Change-Id: I55ea9863a7b6a3aab406e3afdc229800f14812e8
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/381324
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/net/ethernet/stmicro/stmmac/stmmac.h         |  1 -
 drivers/net/ethernet/stmicro/stmmac/stmmac_ethtool.c | 10 ++--------
 drivers/net/ethernet/stmicro/stmmac/stmmac_main.c    |  1 -
 3 files changed, 2 insertions(+), 10 deletions(-)

--- a/drivers/net/ethernet/stmicro/stmmac/stmmac.h
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac.h
@@ -247,7 +247,6 @@ struct stmmac_priv {
 	u32 msg_enable;
 	int wolopts;
 	int wol_irq;
-	bool wol_irq_disabled;
 	int clk_csr;
 	struct timer_list eee_ctrl_timer;
 	int lpi_irq;
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_ethtool.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_ethtool.c
@@ -761,16 +761,10 @@ static int stmmac_set_wol(struct net_dev
 	if (wol->wolopts) {
 		pr_info("stmmac: wakeup enable\n");
 		device_set_wakeup_enable(priv->device, 1);
-		/* Avoid unbalanced enable_irq_wake calls */
-		if (priv->wol_irq_disabled)
-			enable_irq_wake(priv->wol_irq);
-		priv->wol_irq_disabled = false;
+		enable_irq_wake(priv->wol_irq);
 	} else {
 		device_set_wakeup_enable(priv->device, 0);
-		/* Avoid unbalanced disable_irq_wake calls */
-		if (!priv->wol_irq_disabled)
-			disable_irq_wake(priv->wol_irq);
-		priv->wol_irq_disabled = true;
+		disable_irq_wake(priv->wol_irq);
 	}
 
 	mutex_lock(&priv->lock);
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -3538,7 +3538,6 @@ static int stmmac_request_irq_multi_msi(
 	/* Request the Wake IRQ in case of another line
 	 * is used for WoL
 	 */
-	priv->wol_irq_disabled = true;
 	if (priv->wol_irq > 0 && priv->wol_irq != dev->irq) {
 		int_name = priv->int_name_wol;
 		sprintf(int_name, "%s:%s", dev->name, "wol");
